Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php

```diff
@@ -234,26 +223,14 @@ public function getCheckpointStorage(): DbalCheckpointStorage
         return $this->checkpointStorage;
     }
 
-    public function getState(): ContentGraph
+    public function getState(): ContentGraphFinder
     {
-        if (!$this->contentGraph) {
-            $this->contentGraph = new ContentGraph(
-                $this->dbalClient,
-                $this->nodeFactory,
-                $this->contentRepositoryId,
-                $this->nodeTypeManager,
-                $this->tableNamePrefix
-            );
-        }
-        return $this->contentGraph;
+        return $this->contentGraphFinder;
     }
 
     public function markStale(): void
     {
-        $contentGraph = $this->getState();
-        foreach ($contentGraph->getSubgraphs() as $subgraph) {
-            $subgraph->inMemoryCache->disable();
-        }
+        $this->getState()->forgetInstances();
     }
```

Comment:
this changes behaviour as in previously an instance of a `$subgraph` held by a cr user would get its cache flushed. Now not anymore.
This is not certainly a dealbreaker but we should most definitely define the behaviour as we expect it either by test and or by documentation so we can just copy and pasta it to slack ;)

Reply:
good point, this _is_ a dealbreaker IMO – probably it makes sense to call this something else (e.g. `resetRuntimeCaches`) and then reset the inMemoryCache (we should not _disable_ it).

The main goal of this change was not to expose all those nasty internals such that we can refactor more easily.
In fact `inMemoryCache` is now only public because it is still used by the nasty code of the [frontend controller](https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Classes/Controller/Frontend/NodeController.php#L314-L333) (which I plan to clean up)

But we should make sure that:
```php
$subGraph->find*(...);
$contentRepository->handle(...);
$subGraph->find*(...);
```
behaves as expected, i.e. the second subgraph interaction contains the changes.

Lets see if the tests fail – and if they don't we should add scenarios that reflect the above!

Reply:
👍 except that the code you showed is doomed in some scenarios either way

```
$subGraph->find*(...);
$contentRepository->handle(DiscardChanges(...));
$subGraph->find*(...);
```

will not work on a more fundamental way, because the content stream id changes and the one bound to the instance of a $subgraph or $contentgraph contains the old one.
Im not sure what the behaviour will be at this point but likely it will just return nothing as the old content stream is dead.

Reply:
We'll have to think about that one, but once we no longer refer content streams it _should_ be possible I think

Reply:
i thought the assumption was that the contentStreamId will be resolved ON construction of a subgraph (or even before that, on construction of a contentGraph and then `getSubgraph` will just forward it).
This is imo an important aspect that we need to figure out.

Basically the question of joining at runtime ALWAYS or caching the csId by setting it on construction.
Especially since there will be `contentGraph::getContentStreamId` this question is important.

But i thought we agreed upon that the subgraph / or content graph will NOT mutate and operate on the state it was created even if this means that it will fail because the cs id is no longer existing.

so this will be the safest, and if we embrace this pattern we wont need to clear the in memory cache ^^

```php
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
$contentRepository->handle(DiscardChanges(...));
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
```

Reply:
Seeing the one broken test says something has chnaged and I think it's somehow related to these caches, although I can't really figure out what yet.

Reply:
> caching the csId by setting it on construction

IMHO we need that. A ContentGraph and by extension it's subgraphs are bound to a workspace/cs combo. I don't see way around that.

Reply:
yes, I think you're right and I mixed up things.
What I meant to say is:
```php
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
$contentRepository->handle(MoveNodeAggregate(...));
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
```
The second call should not just return the runtime-cached results.

For the (more unlikely) case that you discard/publish/rebase during one process it would be fine to require re-fetching of the subgraph IMO – but the old instance should not just return previously cached results in that case either

Reply:
Indeed!

Reply:
i discussed this matter again with Bernhard and we agree that having a subgraph instance in your hand after doing major changes like Discard or Publish it must be refetched as otherwise the behaviour is undefined (we werent even sure if the old cs-id still exists then in the graph projection)

BUT for simple changes like setting a property a cr user will expect the in memory cache to be flushed (not disabled!).

Reply:
ups i was trying to add it here but then in struck me yes no sure that wont work because we dont know about this implementation detail and would have to hack:

```php
public function forgetInstances(): void
{
    // we also clear the in memory caches for the case someone still holds an instance to a $subgraph
    // and the subgraph should continue to work when fetching an updated node after changing its properties.
    // if the workspace to content stream relation changes using the subgraph or content graph is undefined behaviour!
    foreach ($this->contentGraphInstances as $contentGraph) {
        if (method_exists($contentGraph, 'getSubgraphs')) {
            foreach ($contentGraph->getSubgraphs() as $subgraph) {
                $subgraph->inMemoryCache->disable();
            }
        }
    }

    $this->contentGraphInstances = [];
}
```

Reply:
This discussion should go to another ticket? Because it's now completely unrelated to this change as the inMemoryCache will be a new instance via markStale() so for write side this is mitigated, and everythign else is another story

Reply:
As discussed we are ok with this regression and fix this in a followup: https://github.com/neos/neos-development-collection/issues/5039